scripts/pop-ci/src/git.rs

```diff
@@ -0,0 +1,152 @@
+use std::{
+    fs,
+    io,
+    path::{Path, PathBuf},
+    process,
+    str,
+};
+
+use crate::util::{check_output, check_status};
+
+#[derive(Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
+pub struct GitBranch(String);
+
+impl GitBranch {
+    pub fn new(id: &str) -> Self {
+        Self(id.to_owned())
+    }
+
+    pub fn id(&self) -> &str {
+        &self.0
+    }
+}
+
+#[derive(Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
+pub struct GitCommit(String);
+
+impl GitCommit {
+    pub fn new(id: &str) -> Self {
+        Self(id.to_owned())
+    }
+
+    pub fn id(&self) -> &str {
+        &self.0
+    }
+}
+
+#[derive(Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
+pub struct GitRemote(String);
+
+impl GitRemote {
+    pub fn new(id: &str) -> Self {
+        Self(id.to_owned())
+    }
+
+    pub fn origin() -> Self {
+        Self::new("origin")
+    }
+
+    pub fn id(&self) -> &str {
+        &self.0
+    }
+}
+
+pub struct GitRepo(pub PathBuf);
+
+impl GitRepo {
+    pub fn new<P: AsRef<Path>>(dir: P) -> io::Result<Self> {
+        fs::canonicalize(dir.as_ref()).map(Self)
+    }
+
+    pub fn path(&self) -> &Path {
+        &self.0
+    }
+
+    pub fn command(&self) -> process::Command {
+        let mut command = process::Command::new("git");
+        command.arg("-C").arg(&self.path());
+        command
+    }
+
+    pub async fn async_fetch(&mut self, remote: &GitRemote) -> io::Result<()> {
+        async_std::process::Command::new("git")
+            .arg("-C").arg(&self.path())
+            .arg("fetch")
+            .arg("--prune")
+            .arg("--quiet")
+            .arg("--")
+            .arg(&remote.id())
+            .status()
+            .await
+            .and_then(check_status)
```

Comment:
`async-std` will spawn a thread for each command being awaited. But we could use `async-pidfd` to await the status of a command without spawning a thread from a normal `std::process::Command`.

```rust
let child = command.spawn()?;
let pidfd = AsyncPidFd::from_pid(child.id())?;
check_status(pidfd.wait().await?.status())
```

Reply:
> `async-std` will spawn a thread for each command being awaited. But we could use `async-pidfd` to await the status of a command without spawning a thread from a normal `std::process::Command`.
> 
> ```rust
> let child = command.spawn()?;
> let pidfd = AsyncPidFd::from_pid(child.id())?;
> check_status(pidfd.wait().await?.status())
> ```

That is a good idea, thanks